mpes/fprocessing.py

```diff
@@ -2297,6 +2455,329 @@ def distributedBinning(self, axes, nbins, ranges, binDict=None, pbar=True,
+                            }
+        # import meta data from data file
+        file0 = self.datafiles[0]
+        if 'file' not in metadata_dict.keys():  # If already present, the value is assumed to be a dictionary
+            metadata_dict['file'] = {}
+        
+        print("Collecting lens voltages etc...")
+        with h5py.File(file0, 'r') as f:
+            for k,v in f.attrs.items():
+                if "VSet" in k:  # Changing VSet to V in dict keys
+                    k = k[:-3]
+                metadata_dict['file'][k] = v
+
+        metadata_dict['entry_identifier'] = self.datafolder[13:-2]
+
+        print("Collecting data from the EPICS archive...")
+        # Get metadata from Epics archive if not present already
+        filestart = dt.datetime.utcfromtimestamp(tsFrom/1000).isoformat()
+        fileend = dt.datetime.utcfromtimestamp(tsTo/1000).isoformat()
+        Epics_channels = ["KTOF:Lens:Extr:I", "trARPES:Carving:TEMP_RBV",
+                        "trARPES:XGS600:PressureAC:P_RD", "KTOF:Lens:UDLD:V", "KTOF:Lens:Sample:V",
+                        "KTOF:Apertures:m1.RBV", "KTOF:Apertures:m2.RBV", "KTOF:Apertures:m3.RBV"] + \
+                        [f"trARPES:Carving:{x}.RBV" for x in ['TRX','TRY','TRZ','THT','PHI','OMG']]
+        
+        channels_missing = set(Epics_channels)-set(metadata_dict['file'].keys())
+        for channel in channels_missing:
+            try:
+                req_str = "http://aa0.fhi-berlin.mpg.de:17668/retrieval/data/getData.json?pv=" + \
+                           channel + "&from=" + filestart + "Z&to=" + fileend + "Z"
+                req = urlopen(req_str)
+                data = json.load(req)
+                if data:
+                    vals = [x['val'] for x in data[0]['data']]
+                    metadata_dict['file'][f'{channel}'] = np.average(np.array(vals))
+                else:
+                    metadata_dict['file'][f'{channel}'] = np.nan
+                    print(f"Data for channel {channel} doesn't exist for time {filestart}")     
+            except urllib.error.HTTPError as e:
+                print(f"Incorrect URL for the archive channel {channel}. "
+                     "Make sure that the channel name and file start and end times are correct.")
+                print("Error code: ", e)
+            except urllib.error.URLError as e:
+                print(f"Cannot access the archive URL for channel {channel}. "
+                      f"Make sure that you are within the FHI network."
+                      f"Skipping over channels {channels_missing}.")
+                print("Error code: ", e)
+                break
+        
+        if "sample" not in metadata_dict.keys():
+            metadata_dict['sample'] = {}
+        if "trARPES:Carving:TEMP_RBV" in metadata_dict['file'].keys() and \
+           "trARPES:XGS600:PressureAC:P_RD" in metadata_dict['file'].keys():

```

Comment:
No need for the .keys(), I think. Also, why do we need these copies anyways? Can't we handle this in the reader configuration?